### PR TITLE
Do not update HepMCProduct in place (producer implementation)

### PR DESCRIPTION
**Pull request by wmtan**
This PR is the chosen implementation for avoiding modifying HepMCProduct in place (aka smearing the vertices in place). The smearers (aka vertex generators) still run as EDProducers, but they copy the HepMCProduct, smear the copy, and put the copy into the event with module label "VtxSmeared".
In order to make this PR backward compatible (see below) another producer with label "generatorSmeared"copies the HepMCProduct, and the HepMCProduct with label "generatorSmeared" is the one that is kept on output.
The consumers of the smeared HepMCProduct are modified to use the module label "generatorSmeared" instead of "generator". Also, the generators now create the HepMCProduct with the product instance name "unsmeared", so consumers of the unsmeared HepMCProduct must explicitly use "unsmeared" as the product instance name, in addition to the "generator" module label. This is a defensive measure to ensure that consumers always get the smeared HepMCProduct unless they explicitly specify "unsmeared".
For backward compatibility, the producer of HepMCProduct with label "generatorSmeared" also runs at the beginning of the digitization step. If it finds a HepMCproduct with label "generator" or "VtxSmeared", it copies it, creating a HepMCProduct with module label "generatorSmeared". In this way, files containing GEN-SIM written prior to this PR can still be read.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by wmtan** [comment]
@cmsbuild @davidlange6 the unit test that fails also fails in the base IB. The failure is not caused by this PR.

**Comment by lveldere** [comment]
+1

fastsim edits look ok,
differences observed in fastsim comparisons,
look compatible with fullsim comparisons,
and appear to be compatible with the expected,
given the change in mc truth definition.

**Comment by davidlange6** [comment]
I would have expected no differences in the comparison given the purpose of this PR. Something is not ok. (but I’m happy to be corrected)

**Comment by lveldere** [comment]
The trackingparticles definition changes,
so everything that depends on truth matching can change.

I admit I've been a bit sloppy,
and I don't know exactly which of e.g. these distributions depend on truth
matching:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-08-19-1100+10858/9549/5.1_TTbar+TTbarFS+HARVESTFS/Tracking_Track_general_trackingParticleRecoAsssociation.html

**Comment by makortel** [comment]
TrackingParticles indeed show some differences in their production vertex (TrackingVertex) positions, in both FullSim
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-08-19-1100+10858/9549/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/Tracking_TrackingMCTruth_TrackingParticle.html
and FastSim
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-08-19-1100+10858/9549/5.1_TTbar+TTbarFS+HARVESTFS/Tracking_TrackingMCTruth_TrackingParticle.html

I don't understand why the TrackingParticles change though (from reading Bill's description and browsing through the diff).
* Before: smeared "generator" HepMCProduct was used
* After: smeared "VtxSmeared" HepMCProduct is used, or if that doesn't exist, in MixingModule read "generator" HepMCProduct (but that should also be smeared as before, right?)

I'd expect the end result to be the same in both cases.

On the other hand, the position distributions of the primary TrackingVertices of the in-time pileup events do not seem to change
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_6_X_2015-08-19-1100+10858/9549/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/Vertexing_PrimaryVertexV.html
(if there would be differences, the page above would show them) Are the changes in TrackingParticles only in the positions of "secondary" TrackingParticles then? (just thinking out loud)

**Comment by makortel** [comment]
@lveldere All the plots you linked depend on TrackingParticles (IIRC `num_reco_*` histograms in MultiTrackValidator are the only one that depend only on reco::Tracks).

**Comment by lveldere** [comment]
@davidlange6 

I take my words back, 
I expect no change.

**Comment by bendavid** [comment]
Hi,
General comment: anything which takes HepMC as input to compare with simulation/reconstruction needs to be using the vtxsmeared collection as input.  The only things which can safely use the unsmeared input are generator filters, which in almost all cases do not use or care about the absolute position of the vertices, and use only kinematic quantities which are invariant under the smearing.

Tracking particles and any related validation modules probably need to use the smeared collection.

**Comment by wmtan** [comment]
I will investigate in detail why the vertex positions appear to be different with and without his PR.
I will do this with relvals 5.1 and 25.0.

**Comment by wmtan** [comment]
I found the bug. It is a preexisting horrible bug. The HepMCProduct constructor from a HepMC::GenEvent left the isVtxGenApplied_ , isVtxBoostApplied_, and isPBoostApplied_ flags uninitialized. So, therefore, some of the time the smearing producer thought the smearing had already been applied, and did not smear the vertex. I will fix this.

**Comment by wmtan** [comment]
The bug has been fixed, and this PR has been updated. Please redo the comparisons.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by wmtan** [comment]
Again, this unit test error is in the base IB and not caused by this PR.

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
@cmsbuild please test
the IB problem is apparently gone by now

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10858	b8b4cd0
- changes in  RECO packages are only in test/example files
- jenkins tests pass and comparisons with baseline show no difference

**Comment by wmtan** [comment]
@diguida, @cerminar, @covarell, @bendavid, @vciulli, @Dr15Jones, @deguio, @vadler, @mmusich, @danduggan, @thuer, @monttj, @franzoni
ping

**Comment by Dr15Jones** [comment]
+1

**Comment by monttj** [comment]
+1

**Comment by mmusich** [comment]
+1

**Comment by wmtan** [comment]
@covarell, @bendavid, @vciulli,  @deguio, @danduggan, @thuer, @franzoni
ping again

**Comment by davidlange6** [comment]
Hi all - please raise any remaining issues with this PR by today. Thanks!

**Comment by wmtan** [comment]
@covarell, @bendavid, @vciulli, @thuer, @franzoni This PR was submitted in its final form 12 days ago.
Please sign, or provide a reason for not signing.

**Comment by bendavid** [comment]
(Sorry, was travelling.)

The code changes for the generator part is ok, but if I understand correctly, some additional work would be needed to support backwards compatibility.

In particular when running DIGI-RECO on old GEN-SIM there is the GenParticleProducer plus any validation module which reads the HepMC event.

If we need backwards compatibility for DIGI-RECO but not for validation, then I think the GenParticleProducer is the only module beyond the mixing module which would need to be able to switch.

If backwards compatibility is needed for validation then it's more complicated.  @davidlange6 any comment?

@wmtan, you said "The backward compatibility issue does not cause any known relval matrix failures."  Do we know exactly which steps those tests run on old GEN-SIM?

**Comment by wmtan** [comment]
@bendavid I ran the entire relval matrix (not just the limited one), and the only backward compatibility issue I found was the mixing module reading old minbias files. So, I updated the mixing module to be fully backward compatible. Since I saw no other failures, I assume that, in the relval matrix, no other old input files in which HepMCProduct needs to be read are usedin the relval matrix tests.
Incidentally, it would be much easier to add backward compatibility to GenParticleProducer and any other modules that need it after this PR is merged.

**Comment by davidlange6** [comment]
Hi Bill

Could you run 

runTheMatrix.py -i all -l 1306 

and confirm it works in your area?  This would recycle the existing GEN-SIM, which perhaps you did not already do.

Thx-
David

**Comment by wmtan** [comment]
@davidlange6 OK, I will do this (runTheMatrix.py -i all -l 1306), but it will take some time in that the IB used for my work area is no longer supported.

**Comment by wmtan** [comment]
@davidlange6 "runTheMatrix.py -i all -l 1306" gives a product not found exception because GenParticleProducer is not backward compatible. Should I update this PR with the changes needed for this backward compatibility, or should that be done with a separate PR? If I update this PR, I think only generators and simulation should need to resign it.

**Comment by davidlange6** [comment]
@wmtan - lets add it to this PR.

**Comment by wmtan** [comment]
The chnages for backward compatibility have been added to this PR.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by deguio** [comment]
+1

**Comment by diguida** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by slava77** [comment]
@cmsbuild please test
(I can't imagine that it takes more than 2 days to test and still expect a successful completion)

**Comment by Dr15Jones** [comment]
please test

**Comment by wmtan** [comment]
@covarell, @bendavid, @govoni, @vciulli, @cvuosalo, @slava77, @vadler, @thuer, @monttj, @franzoni ping: analysis, generators, operations, reconstruction Can you either sign this, or provide a reason for not signing.

File: GeneratorInterface/Core/interface/GeneratorSmearedProducer.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace edm {
+  class ParameterSet;
+  class ConfigurationDescriptions;
+  class Event;
+  class EventSetup;
+  class HepMCProduct;
+}
+
+class GeneratorSmearedProducer : public edm::EDProducer {
```
Comment (slava77) [inline GeneratorInterface/Core/interface/GeneratorSmearedProducer.h]: why is this one not thread-aware? (edm::one, stream )

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/10858	 f719245
- code changes are mostly as expected
- a newly introduced  producer ``` class GeneratorSmearedProducer : public edm::EDProducer ``` should be changed to an edm::one, I suppose, but better do it in a separate PR to have less signatures to collect
- jenkins tests pass and comparisons with the baseline show no differences

File: GeneratorInterface/Core/interface/GeneratorSmearedProducer.h (hunk 1 above)
Comment (wmtan) [inline GeneratorInterface/Core/interface/GeneratorSmearedProducer.h]: Yes, this new producer, as well as the modified vertex smeared producer, should be thread aware (e.g. one, stream, or global), just as all other modules should be. That should happen in due course.

**Comment by wmtan** [comment]
@covarell, @bendavid, @govoni, @vciulli, @vadler, @thuer, @monttj, @franzoni 
ping again: analysis, generators, operations Can you either sign this, or provide a reason for not signing?

**Comment by ferencek** [comment]
@wmtan, this PR has broken the jet flavor code. In CMSSW_7_6_X_2015-10-07-2300 I now get a bunch of warnings like the following one
```
Begin processing the 500th record. Run 1, Event 1800, LumiSection 18 at 08-Oct-2015 00:32:46.939 CDT
%MSG-w UndefinedPartonMode:  HadronAndPartonSelector:patJetPartons  08-Oct-2015 00:32:46 CDT Run: 1 Event: 1800
Could not automatically determine the hadronizer type and set the correct parton selection mode. Parton-based jet flavour will not be defined.
```
when running PAT tests. How did this go unnoticed?

Anyways, the hadron and parton selector defined in `PhysicsTools/JetMCAlgos/plugins/HadronAndPartonSelector.cc` needs to access the `GenEventInfoProduct` in order to determine the hadronizer type. What is the proposed solution here. Is that product still available in the event and under what name?

**Comment by wmtan** [comment]
@ferencek This PR should not have affected the GenEventInfoProduct in any way.
Please provide evidence that this PR broke the jet flavor code.
At a minimum, I would need a cmsRun job that reproduces this error.
I cannot provide a solution to a problem that I do not understand, and that, at least on the surface, appears to have nothing whatever to do with this PR.

**Comment by ferencek** [comment]
@wmtan, evidence is the printout I provided above. The problem can be reproduced by runnning pretty much any of the PAT unit tests. For instance, running `PhysicsTools/PatAlgos/test/patTuple_standard_cfg.py` should suffice.

The problem originates from the change made in ` PhysicsTools/PatAlgos/python/mcMatchLayer0/jetFlavourId_cff.py`. If the `GenEventInfoProduct` is unaffected, then changing
```python
src = cms.InputTag("generatorSmeared"),
```
back to
```python
src = cms.InputTag("generator"),
```
should fix the issue.

**Comment by wmtan** [comment]
1) The reason this problem was not found is that it does not cause any tests to fail. The PR approval process does not scan all output files for new warning messages.
2) I now understand the problem. It is only a backward compatibility issue. GenEventInfoProduct is indeed unaffected. Changing "generatorSmeared" back to "generator" does indeed fix the unit test, but only because the file being read predates this PR. Once a new input file for this test is generated, it would have to be changed back to "generatorSmeared".
3) Backward compatibility was indeed addressed in this PR, but the fix does not cover the unusual workflow in this test.
4) Further discussion should be in e-mail outside of github. If you want a record of the further discussion, use an appropriate HN forum.

**Comment by ferencek** [comment]
@wmtan, ok, it sounds like the problem will be fixed once the input RelVals are updated to a version that contains this PR. In that case, I don't think any fixes are needed.

**Comment by wmtan** [comment]
@ferencek That is correct, assuming that the relvals in question smear the vertices in HepMCProduct, which they should.

**Comment by ferencek** [comment]
@wmtan, I had a look at one of the files from /RelValProdTTbar_13/CMSSW_7_6_0_pre7-76X_mcRun2_asymptotic_v5-v1/AODSIM and here is what I see
```
[ferencek@cmslpc25 src]$ edmDumpEventContent root://cmsxrootd-site.fnal.gov//store/relval/CMSSW_7_6_0_pre7/RelValProdTTbar_13/AODSIM/76X_mcRun2_asymptotic_v5-v1/00000/0E9A5DE8-1D71-E511-A205-00261894380D.root
Type                                  Module                      Label             Process   
----------------------------------------------------------------------------------------------
GenEventInfoProduct                   "generator"                 ""                "SIM"     
edm::TriggerResults                   "TriggerResults"            ""                "SIM"     
edm::TriggerResults                   "TriggerResults"            ""                "HLT"     
int                                   "addPileupInfo"             "bunchSpacing"    "HLT"     
vector<PileupSummaryInfo>             "addPileupInfo"             ""                "HLT"     
vector<int>                           "genParticles"              ""                "HLT"
...
```
and
```
[ferencek@cmslpc25 src]$ edmProvDump  root://cmsxrootd-site.fnal.gov//store/relval/CMSSW_7_6_0_pre7/RelValProdTTbar_13/AODSIM/76X_mcRun2_asymptotic_v5-v1/00000/0E9A5DE8-1D71-E511-A205-00261894380D.root | grep -i cmssw
  SIM '' '"CMSSW_7_6_0_pre7"' (6459b08dbcc3c6e5421bfc57cfa6cb04)
    HLT '' '"CMSSW_7_6_0_pre7"' (e355911c26dddb8d43fe6d5bca5bbf21)
      RECO '' '"CMSSW_7_6_0_pre7"' (39f973222d399b2f7e7caed12b632381)
  EfficiencyMap: string tracked  = 'CondFormats/JetMETObjects/data/CMSSW_538_TrackNonEff.txt'
  LeakageMap: string tracked  = 'CondFormats/JetMETObjects/data/CMSSW_538_TrackLeakage.txt'
  ResponseMap: string tracked  = 'CondFormats/JetMETObjects/data/CMSSW_538_response.txt'
  tagName: vstring tracked  = {'ZSP_CMSSW390_Akt_05_PU0'}
   FileName: FileInPath tracked  = V001 SimG4CMS/Forward/data/CastorShowerLibrary_CMSSW500_Standard.root 2 /external/slc6_amd64_gcc493/data/SimG4CMS/Forward/data/CastorShowerLibrary_CMSSW500_Standard.root
```
and I don't see any GenEventInfoProducts called "generatorSmeared". I'm not sure if that was expected but if so, then the changes made in `PhysicsTools/PatAlgos/python/mcMatchLayer0/jetFlavourId_cff.py` should be reverted.

**Comment by wmtan** [comment]
@ferencek The label "generatorSmeared" only applies to HepMCProduct, not to GenEventInfoProduct.
So, nothing is wrong here.

**Comment by ferencek** [comment]
@wmtan, thanks for the reply. Then my comment stands. The change done in `PhysicsTools/PatAlgos/python/mcMatchLayer0/jetFlavourId_cff.py` should be reverted since the modified parameter does not refer to a HepMCProduct but a GenEventInfoProduct. I can make a PR for that if you agree that this is what should be done. Thanks

**Comment by wmtan** [comment]
@ferencek Yes, you are right. It does refer to a GenEventInfoProduct. The change to that one file should be reverted.

**Pull request merged**